lib/utils/report.mjs

```diff
@@ -164,39 +175,142 @@ function isDisabled(ruleName, startLine, disabledRanges) {
 	return false;
 }
 
-/** @param {Problem & { line: number }} problem */
-function isFixApplied({ fix, line, result: { stylelint }, ruleName }) {
-	const { disabledRanges, config = {}, fixersData } = stylelint;
+/**
+ * @param {Problem & { line: number }} problem
+ * @returns {{range: PostcssRange, text: string, applied: false} | {applied: boolean}}
+ */
+function applyFix({ fix, fixedNode, line, result: { stylelint }, ruleName }) {
+	const { disabledRanges, config = {}, fixersData, fixedRanges } = stylelint;
+
+	if (!isFn(fix)) return { applied: false };
 
-	if (!isFn(fix)) {
-		addFixData({ fixersData, ruleName, fixed: false });
+	const shouldFix = Boolean(config.fix || (config.computeReplacementText && fixedNode));
+	const shouldFixThisRule = Boolean(shouldFix && !config.rules?.[ruleName][1]?.disableFix);
+	const mayFix =
+		shouldFixThisRule && (config.ignoreDisables || !isDisabled(ruleName, line, disabledRanges));
+
+	if (!mayFix) return { applied: false };
+
+	fix();
+
+	if (!config.computeReplacementText) {
+		addFixData({ fixersData, ruleName });
 
-		return false;
+		return { applied: true };
 	}
 
-	const shouldFix = Boolean(config.fix && !config.rules?.[ruleName][1]?.disableFix);
-	const mayFix =
-		shouldFix && (config.ignoreDisables || !isDisabled(ruleName, line, disabledRanges));
+	if (!fixedNode?.source?.start || !fixedNode?.source?.end) {
+		return { applied: false };
+	}
 
-	addFixData({ fixersData, ruleName, fixed: mayFix });
+	const fixedNodeRange = {
+		start: fixedNode.source.start,
+		end: fixedNode.source.end,
+	};
 
-	if (!mayFix) return false;
+	// When recording replacement text we want to ensure that there is no overlap with any other fix.
+	// We only record the first fix for each node.
+	if (fixedRanges.some((fixedRange) => rangesOverlap(fixedRange, fixedNodeRange)))
```

Comment:
question

This implies that we never reset the initial `fixData.text` for all overlaps?
i.e. `fixedNode.toString()` is called once for that range
Wouldn't last wins be more logical?
That would require some refactors in the case that the node is different though.
e.g. sibling, child, etc.

Reply:
Why would we want the last one?
Then we would have bleed between fixers.

This is the entire concept of this change and what makes this feasible :)

By marking entire ranges as "modified" and ignoring them in subsequent `report()` calls we are sure that we only recorded a single fix from a single rule for a single region in the source code.

We must not have any overlap.

If however we constantly override we would have combined fixes.
Thereby the end user would lose all granularity.

They would receive a single `replacement text` containing the combined mutations of several rules. But all squished together in a warning for what happened to have been the last rule to run.

Reply:
> and _for some_ they will have the computed replacement text

That part will have to be documented so that user will have why it's missing.

Reply:
Yes, how `replacement text` is calculated and the limitations will have to be documented.

Also see how eslint describes the same challenge: https://eslint.org/docs/latest/extend/custom-rules#conflicting-fixes